File: vlq.js
const fs = require('fs');
const path = require('path');

module.exports.createDecoder = function createDecoder(delegate) {
  return WebAssembly.compile(fs.readFileSync(path.join(__dirname, 'vlq.wasm')))
    .then((mod) => {
      return WebAssembly.instantiate(mod, {
        env: {
          emitNewline() {
            delegate.emitNewline();
          },
          emitMapping1(col) {
            delegate.emitMapping1(col);
          },
          emitMapping4(col, src, srcLine, srcCol) {
            delegate.emitMapping4(col, src, srcLine, srcCol);
          },
          emitMapping5(col, src, srcLine, srcCol, name) {
            delegate.emitMapping4(col, src, srcLine, srcCol, name);
          },
        }
      });
    })
    .then((instance) => new Decoder(instance));
}

class Decoder {
  constructor(mod) {
    const {
      decodeVLQ,
      decode,
      memory,
    } = mod.exports;
    this._mod = mod;
    this._memory = memory;
    this._decode = decode;
    this._decodeVLQ = decodeVLQ;
  }

  decode(str) {
    this._writeReader(512, str);
    this._decode(512);
  }

  decodeVLQ(str) {
    this._writeReader(512, str);
    return this._decodeVLQ(512);
  }

  _ensureMem(str) {
    const needed = str.length - this._memory.buffer.byteLength + 1024;
    if (needed > 0) {
      const pages = Math.ceil(needed / 65536);
      this._memory.grow(pages);
    }
  }

  /**
   * write struct Reader
   * @param {number} ptr
   * @param {string} str
   */
  _writeReader(ptr, str) {
    this._ensureMem(str);

    const stringPtr = ptr + 16;
    const heap32 = new Int32Array(this._memory.buffer);
    heap32[(ptr >> 2)] = 0;
    heap32[(ptr >> 2) + 1] = stringPtr;
    heap32[(ptr >> 2) + 2] = str.length;
    this._writeString(stringPtr, str);
  }

  /**
   * Write ascii string to memory
   * @param {Uint8Array} heap8
   * @param {number} ptr
   * @param {string} str
   */
  _writeString(ptr, str) {
    const bytes = new Uint8Array(this._memory.buffer);
    for (let i = 0; i < str.length; i++) {
      bytes[ptr + i] = str.charCodeAt(i);
    }
  }
}

File: vlq.c
#include "vlq.h"

unsigned char asciiToUint6[127] = {
  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,62,0,0,0,63,52,53,54,55,56,57,58,59,60,61,0,0,0,0,0,0,0,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,0,0,0,0,0,0,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,0,0,0,0
};

int decodeVLQ(Reader *reader) {
  int num = 0;
  int shift = 0;
  int digit = 0;
  int cont = 0;
  int negate = 0;
  do {
    digit  = asciiToUint6[reader->bytes[reader->pos++] & 0x7F];
    cont   = digit & 32;
    digit  = digit & 31;
    if (shift == 0) {
      negate = digit & 1;
      num = digit >> 1;
      shift = 4;
    } else {
      num |= digit << shift;
      shift += 5;
    }
  } while (cont && shift < 30);
  return negate ? -num : num;
}

void emitMapping(int fieldCount, int column, int source, int sourceLine, int sourceColumn, int name) {
  switch (fieldCount) {
    case 1:
      emitMapping1(column);
      break;
    case 4:
      emitMapping4(column, source, sourceLine, sourceColumn);
      break;
    case 5:
      emitMapping5(column, source, sourceLine, sourceColumn, name);
      break;
  }
}

void decode(Reader *reader) {
  unsigned char byte;
  int fieldCount = 0;
  int line = 0;
  int column = 0;
  int source = 0;
  int sourceLine = 0;
  int sourceColumn = 0;
  int name = 0;
  int value = 0;

  while ((byte = reader->bytes[reader->pos]) != 0) {
    switch (byte) {
      case 59: // semicolon
        if (fieldCount > 0) {
          emitMapping(fieldCount, column, source, sourceLine, sourceColumn, name);
        }
        emitNewline();
        column = 0;
        fieldCount = 0;
        reader->pos++;
        break;
      case 44: // comma
        emitMapping(fieldCount, column, source, sourceLine, sourceColumn, name);
        fieldCount = 0;
        reader->pos++;
        break;
      default:
        value = decodeVLQ(reader);
        switch (fieldCount) {
          case 0:
            column += value;
            fieldCount = 1;
            break;
          case 1:
            source += value;
            fieldCount = 2;
            break;
          case 2:
            sourceLine += value;
            fieldCount = 3;
            break;
          case 3:
            sourceColumn += value;
            fieldCount = 4;
            break;
          case 4:
            name += value;
            fieldCount = 5;
            break;
        }
        break;
    }
  }
  if (fieldCount > 0) {
    emitMapping(fieldCount, column, source, sourceLine, sourceColumn, name);
  }
}
